src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Monitor.NativeAot.cs

```diff
@@ -240,13 +240,25 @@ private static object CreateThreadLocalContentionCountObject()
         {
             Debug.Assert(t_ContentionCountObject == null);
 
+            // we use "s_lockContentionCounter" as a sentinel value to prevent reentrancy.
+            t_ContentionCountObject = s_lockContentionCounter;
+
             object threadLocalContentionCountObject = s_lockContentionCounter.CreateThreadLocalCountObject();
             t_ContentionCountObject = threadLocalContentionCountObject;
             return threadLocalContentionCountObject;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static void IncrementLockContentionCount() => ThreadInt64PersistentCounter.Increment(t_ContentionCountObject ?? CreateThreadLocalContentionCountObject());
+        internal static void IncrementLockContentionCount()
+        {
+            // CreateThreadLocalCountObject may indirectly take locks and in rare cases may have contentions.
```

Comment:
What is the stacktrace of the re-entrancy? `ThreadInt64PersistentCounter` should be using LowLevelLock that should be calling back here.

Reply:
`TypeSystem_LockFreeReaderHashtable` is not actually lock-free. It takes locks and that can cause reentrancy.

Reply:
Hmm, that is pretty ugly re-entrancy. There is also going to be a problem with `RuntimeTypeUnifier__GetRuntimeTypeBypassCache`.

I think we should fix this by preventing this re-entrancy from happening in the first place. Should we switch `LockFreeReaderHashtable` to use low-level lock to prevent this re-entrancy from happening?

Reply:
`ThreadInt64PersistentCounter` could be much simpler. It could be just a striped counter. No need to attach a counter to every thread, etc... 

The root cause is though the fact that type system takes locks, so it makes it fragile when lock implementation needs to touch the type system.

I will take a look at low-level lock.

Reply:
I have tried using LowLevelLock and I see things locking up with roughly the same frequency. But that is not due to reentrancy. It seems more like a bug in LowLevelLock where I see bunch of threads waiting, but nooone owning the lock.

Also LowLevelLock seems too heavy for this purpose. It tries to be fair, registers waiters, and such. Do we have something like a simple critical section?

Reply:
How typical lockup looks like:

![image](https://user-images.githubusercontent.com/8218165/172882658-c2e6752a-9a99-4d16-a388-632d7ad02e43.png)

Reply:
> It seems more like a bug in LowLevelLock where I see bunch of threads waiting, but noone owning the lock.

We use LowLevelLock in regular CoreCLR as well. I am wondering why we are not hitting this bug there.

Reply:
Maybe not enough stress of contentious scenarios. I see no tests directly for LowLevelLock and only a few uses. 

At a first glance all uses of `LowLevelLock` seem to be for simple mutual exclusion, and perhaps all should just use `LowLevelMonitor` or some other critical-section-like thing. When using a lock one typically does not expect contention and if contentions happen does not want fairness.

Reply:
This also makes me think that the lock in `LockFreeReaderHashtable` can be contentious. 
Perhaps only at startup. Since items cannot be removed, I guess, it is mostly read use at steady state.

Reply:
> We use LowLevelLock in regular CoreCLR as well. I am wondering why we are not hitting this bug there.

Now that I have seen more ways how this deadlock may show up, I think we probably always had some circular dependency and it is just harder to untangle when multiple threads are involved. It might not be a problem with LowLevelLock after all. 

I am still not sure why `LowLevelLock` is not just a copy of NativeAOT `Lock` without Wait/Pulse support. Something to consider another day.